### PR TITLE
Feature/jf roku 404 record livetv

### PR DESCRIPTION
**Pull request by jimdogx**
Adds "Record" / "Record Series" buttons to the **TV Guide**

**Changes**
1. Adds two new buttons to the Program Details view of the **TV Guide**.
2. New red circle icon to indicate a show is recording / scheduled to record.

**Issues**
Address some of #404 as it does not add the ability to record from other places (such as while watching a program). I plan on looking in to adding that as well, but thought it might be good to get at least some of the functionality to users sooner rather than later.

**Known Issues**
1. For some reason, the Roku grid does not always _remove_ the red icon when you cancel a recording (the recording is cancelled and buttons are updated correctly though).  Once a user leaves the guide and returns, it refreshes correctly.

2. I also noticed that the Roku grid does not always show the red icon on shows upon loading until a user scrolls to that channel.

3. Currently using the same red circle regardless if it's a one off recording or a series recording.

### CODE REVIEW

**Comment by neilsb** [comment]
This is looking great.   Just had a quick play with it, and scheduled some recordings.  I'll have a further play and proper looks at the code later on.   

Just a few things I've noticed:
 - Added a comment about button presses
 - Web web app doesn't show the "Record Series" for some items, such as Movies.  Here it's always being shown.  I think the web app is looking for the `IsSeries` property in the response to the  `LiveTv/Programs/` call and using that to decide if to show the series record or not.
 - There is something weird going on with not always showing the correct buttons (i.e. Record or Cancel Recording).  Probably related to your point number 1 in known issues, but I've not been able to see what is happening consistently, so can't detail it for you yet.  I'll have a further play, but definitely seeing some "Cancel Recoding" when the recording has already been cancelled.  Exiting and going back in seems to fix it, but can get messed up again when you set other items to record.
 - When you're highlighting an item in the grid that's set to record, the red circle turns black.  Not critical, but would be nice if that could remain red.

File: components/liveTv/ProgramDetails.brs [hunk 1]
```diff
@@ -221,12 +299,40 @@ end sub
...
+        if key = "right" and m.viewChannelButton.hasFocus()
+            m.recordButton.setFocus(true)
+            m.viewChannelOutline.visible = false
+            m.recordOutline.visible = true
+            return true
+        else if key = "right" and m.recordButton.hasFocus()
+            m.recordSeriesButton.setFocus(true)
+            m.recordOutline.visible = false
+            m.recordSeriesOutline.visible = true
+        else if key = "left" and m.recordSeriesButton.hasFocus()
```
Comment (neilsb) [inline components/liveTv/ProgramDetails.brs]: When you're on the Record Button, and press right, the highlighted button moves, but the "not a valid button press" sound is played.

Just missing the `return true` here to let the OS know the event has been handled.

Same is required on the `else if` below

Comment (jimdogx) [reply components/liveTv/ProgramDetails.brs]: Fixed 👍

**Comment by jimdogx** [comment]
- Added a comment about button presses
✅  Done

- Web web app doesn't show the "Record Series" for some items, such as Movies. Here it's always being shown. I think the web app is looking for the IsSeries property in the response to the LiveTv/Programs/ call and using that to decide if to show the series record or not.
✅. Done

- There is something weird going on with not always showing the correct buttons (i.e. Record or Cancel Recording). Probably related to your point number 1 in known issues, but I've not been able to see what is happening consistently, so can't detail it for you yet. I'll have a further play, but definitely seeing some "Cancel Recoding" when the recording has already been cancelled. Exiting and going back in seems to fix it, but can get messed up again when you set other items to record.
📌  Haven't been able to reproduce this, but I'll keep testing it.

- When you're highlighting an item in the grid that's set to record, the red circle turns black. Not critical, but would be nice if that could remain red.
😔  I spent quite a bit of time trying to get this to work.  At one point I even got this: 

Screen Shot 2022-01-10 at 6 51 02 PM

But it is completely random whether Roku draws the png on top of the blue background or blends it:

Screen Shot 2022-01-10 at 6 52 51 PM

**Comment by neilsb** [comment]
It looks like adding the property `iconFocusedColor="#ff0000"` to the `TimeGrid` element fixes the issue with the dot showing as black.

**Comment by neilsb** [comment]
And I think there might only be once case when things get a little odd.  If you go into the guide and mark something as `Record`, and then cancel the recoding, the record icon doesn't vanish, the the buttons still show the cancel the recording.

Going out and back into the guide again shows the item as not being recorded.  Likewise, if you load the guide and there is already something set to record, and you cancel the recording, that all works as expected.  So it's just setting something to record, and then cancel that recording without exiting the guide when it goes wrong.

File: components/liveTv/schedule.brs [hunk 2]
```diff
@@ -180,6 +182,44 @@ sub onWatchChannelSelected()
...
+
+    m.RecordProgramTask = createObject("roSGNode", "RecordProgramTask")
+    m.RecordProgramTask.programDetails = m.detailsPane.programDetails
+    m.RecordProgramTask.recordSeries = true
+    m.RecordProgramTask.observeField("recordOperationDone", "onRecordOperationDone")
+    m.RecordProgramTask.control = "RUN"
+end sub
+
+sub onRecordOperationDone()
+    if m.LoadScheduleTask.state <> "run"
```
Comment (neilsb) [inline components/liveTv/schedule.brs]: Do we need to reload the schedule if it's only a program being recorded?  Understandable if it's a series, but for a single program can we just update the selected program details?

**Review by neilsb — COMMENTED** [review]
I like the outline you've done for the button on the Program Details view.

My one though about that (and this is very much up for discussion) is that maybe we want to use the "Blue" as the "Selected Item" colour.  So rather than having all the buttons as that colour, the focused one is blue and the other buttons are another colour that shows they're buttons, but not focused.

**Comment by jimdogx** [comment]
> It looks like adding the property `iconFocusedColor="#ff0000"` to the `TimeGrid` element fixes the issue with the dot showing as black.

Would love to know how you found that.  I searched all of the Roku Brightscript docs for "iconFocusedColor" and it is no where to be found.  But it works 👍 

>  If you go into the guide and mark something as Record, and then cancel the recoding, the record icon doesn't vanish, the the buttons still show the cancel the recording.

I pushed an updated that appears to work 99% of the time.  I have one channel that when I scroll one time slot to the right, Roku moves the focussed program back one time slot (e.g. I move to 9 - 10 am and it auto moves back to the previous 8 - 9 am slot)!!!  On that channel, the buttons don't get updated correctly.  On all other channels I've tested it works as you'd expect.

> So rather than having all the buttons as that colour, the focused one is blue and the other buttons are another colour that shows they're buttons, but not focused.

Agreed.  Done 👍 

> Do we need to reload the schedule if it's only a program being recorded? Understandable if it's a series, but for a single program can we just update the selected program details?

No we don't and even better, this is what fixed the buttons not getting updated correctly (except for the one channel I mentioned above).

File: components/liveTv/schedule.brs (hunk 2 above)
Comment (jimdogx) [reply components/liveTv/schedule.brs]: Agreed.  See comments below.

**Comment by neilsb** [comment]
> > It looks like adding the property `iconFocusedColor="#ff0000"` to the `TimeGrid` element fixes the issue with the dot showing as black.

> Would love to know how you found that. I searched all of the Roku Brightscript docs for "iconFocusedColor" and it is no where to be found. But it works 👍

I got frustrated and broke out the [RALE tool](http://devtools.web.roku.com/#rale-tool) and started playing with the all the (undocumented) properties until one made the difference.  Just shout on Matrix if you want to chat about the RALE tool at any point.

Saw the button changes, and they look great.  I'll give the whole thing another going one evening this week and get it merged in.  Really great work on this.  👍

**Review by neilsb — COMMENTED** [review]
All looking good now.  Excellent.

**Review by neilsb — APPROVED** [review]

**Pull request merged**